### DISCUSSION TITLE
MCP2515 Config in Modul-Parameters Generic(18)

### DISCUSSION THREAD

**Discussion by reset12** [arendst/Tasmota#18084] [Q&A]
Hello,
i have generated Tasmota-bin with tasmocompiler and defined "define USE_CANSNIFFER " in Config.
I have connected MCP2515 to Wemos D1 mini as following: ESP GPIO4(D2) to MCP2515 INT, ESP GPIO14(D5) to MCP2515 SCK,
ESP GPIO12(D6) to MCP2515 SO, ESP GPIO13(D7) to MCP2515 SI, ESP GPIO15(D8) to MCP2515 CS.

Which Parameters i must select to get my connections work ?

**Comment by barbudor** [comment]
You wrote 'connected' but did you actually configured also those gpio in Configuration => Conigure Module?

**Reply by reset12** [reply]
That is my Problem i need the configuration in Configure Module.
I have physically connected.

**Reply by barbudor** [reply]
There are those lines in the code of the driver:
 ```
 * CAN sniffer using MCP2515 - Microchip CAN controller
 *
 * Connections:
 * MCP2515  ESP8266         Tasmota
 * -------  --------------  ----------
 *  INT     not used
 *  SCK     GPIO14          SPI CLK
 *  SI      GPIO13          SPI MOSI
 *  SO      GPIO12          SPI MISO
 *  CS      GPIO0..5,15,16  MCP2515
 *  Gnd     Gnd
 *  VCC     Vin/5V
 ```

**Comment by reset12** [comment]
Thank you for your Help

**Comment by reset12** [comment]
Hello barbudor,
it doesnt work. 
The console shows:
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F5, Count 15
00:00:00.055 QPC: Count 1
00:00:00.061 SPI: Software using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.075 Project tasmota - Tasmota-CAN Version 12.4.0(tasmota)-2_7_4_9(2023-03-01T18:34:21)
00:00:00.564 WIF: Connecting to AP1 7050-box Channel 6 BSSId E0:28:6D:15:C4:1D in mode 11n as tasmota-B5DF95-8085...
00:00:01.754 WIF: Connected
00:00:02.008 HTP: Web server active on tasmota-B5DF95-8085 with IP address 192.168.178.48
19:57:42.018 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"12.4.0(tasmota)","FallbackTopic":"cmnd/DVES_B5DF95_fb/","GroupTopic":"cmnd/tasmotas/"}}
19:57:42.022 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-B5DF95-8085","IPAddress":"192.168.178.48"}}
19:57:42.033 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":7}}
19:57:45.401 QPC: Reset
19:57:46.453 RSL: STATE = {"Time":"2023-03-01T19:57:46","Uptime":"0T00:00:09","UptimeSec":9,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"7050-box","BSSId":"E0:28:6D:15:C4:1D","Channel":6,"Mode":"11n","RSSI":70,"Signal":-65,"LinkCount":1,"Downtime":"0T00:00:03"}}
19:58:33.783 RSL: RESULT = {"Timers":"OFF","Timer1":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":3},"Timer2":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer3":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer4":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer5":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer6":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer7":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer8":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer9":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer10":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer11":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer12":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer13":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer14":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer15":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0},"Timer16":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Action":0}}
20:02:46.396 RSL: STATE = {"Time":"2023-03-01T20:02:46","Uptime":"0T00:05:09","UptimeSec":309,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"7050-box","BSSId":"E0:28:6D:15:C4:1D","Channel":6,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}

and noting else.

My configuration is:
#define USE_MCP2515 
#define USE_CANSNIFFER

and then modul config i use is:
![main](https://user-images.githubusercontent.com/14790813/222240177-8a70c6dd-7107-4934-bb1b-b090e3e7e89f.png)
![controlbox](https://user-images.githubusercontent.com/14790813/222240657-f106f33e-fea9-41f8-8f79-94053538cf56.jpg)

I like to use Tasmota to control a Powersupply.
Looks the Tasmotascreen Ok and have i to use script or rules to get it work ?

**Reply by barbudor** [reply]
> My configuration is:
#define USE_MCP2515
#define USE_CANSNIFFER

You should include one or the other
I don't know the difference but they are exclusive

**Comment by reset12** [comment]
Ok, i tried only one.

**Comment by reset12** [comment]
I use now only #define USE_MCP2515, but nothing changed.
Is the main Screen Ok or should it display any values ?

**Reply by barbudor** [reply]
Please provide the output of
```
Backlog Template; Module; GPIO 255; status 0
```

**Comment by reset12** [comment]
21:23:43.104 CMD: Backlog Template
21:23:43.130 RSL: RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
21:24:00.112 CMD: Module
21:24:00.120 RSL: RESULT = {"Module":{"18":"Generic"}}
21:24:12.955 CMD: GPIO 255
21:24:12.967 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"832":"SSPI MISO"},"GPIO13":{"864":"SSPI MOSI"},"GPIO14":{"896":"SSPI SCLK"},"GPIO15":{"928":"SSPI CS"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
21:24:24.327 CMD: status 0
21:24:24.338 RSL: STATUS = {"Status":{"Module":18,"DeviceName":"Tasmota-CAN","FriendlyName":["Tasmota-CAN"],"Topic":"tasmota_B5DF95","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
21:24:24.370 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:48:15","StartupUTC":"2023-03-01T19:36:09","Sleep":50,"CfgHolder":4617,"BootCount":5,"BCResetTime":"2023-03-01T20:31:02","SaveCount":14,"SaveAddress":"F6000"}}
21:24:24.403 RSL: STATUS2 = {"StatusFWR":{"Version":"12.4.0(tasmota)","BuildDateTime":"2023-03-01T19:24:32","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"352/699"}}
21:24:24.423 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":4,"LogHost":"","LogPort":514,"SSId":["7050-box",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
21:24:24.454 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":632,"Free":368,"Heap":24,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
21:24:24.485 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-B5DF95-8085","IPAddress":"192.168.178.48","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.178.1","DNSServer2":"0.0.0.0","Mac":"7C:87:CE:B5:DF:95","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
21:24:24.510 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_B5DF95","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
21:24:24.532 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-03-01T20:24:24","Local":"2023-03-01T21:24:24","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"07:32","Sunset":"18:32"}}
21:24:24.550 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-03-01T21:24:24"}}
21:24:24.556 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-03-01T21:24:24","Uptime":"0T00:48:15","UptimeSec":2895,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"7050-box","BSSId":"E0:28:6D:15:C4:1D","Channel":6,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:03"}}}

**Reply by barbudor** [reply]
You have configured SSPI and not SPI
And I don't see any gpio set to MCP2515_CS 


I'm still not sure what this driver is supposed to do
May be the author @mbezuidenhout could provide more details

**Reply by mbezuidenhout** [reply]
To get the debug data you can move line 269
`AddLog(LOG_LEVEL_DEBUG, PSTR("MCP2515: Received message 0x%s from ID 0x%x"), canMsg, (uint32_t)canFrame.can_id);`
to above line 159
`#ifdef MCP2515_BMS_CLIENT`
then you can if will output the received messages in the Console.

The purpose of this driver was to expand it to incorporate various devices that communicate via CAN-bus. At the moment it only contains the code for the FreedomWon batteries.

**Comment by reset12** [comment]
Hello Marius,
in which file i must move the lines ?
I need the CAN-Interface to control my Huawei R4850G2 Power Supply, for adapt the Power delivered into my LIFEPO4 Battery depending on generated Solar-Power.



On Thu, 02 Mar 2023 03:39:58 -0800
Marius Bezuidenhout ***@***.***> wrote:

> To get the debug data you can move line 269
> `AddLog(LOG_LEVEL_DEBUG, PSTR("MCP2515: Received message 0x%s from ID 0x%x"), canMsg, (uint32_t)canFrame.can_id);`
> to above line 159
> `#ifdef MCP2515_BMS_CLIENT`
> then you can if will output the received messages in the Console.
>
> The purpose of this driver was to expand it to incorporate various devices that communicate via CAN-bus. At the moment it only contains the code for the FreedomWon batteries.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/arendst/Tasmota/discussions/18084#discussioncomment-5177924
> You are receiving this because you authored the thread.
>
> Message ID: ***@***.***>


--
***@***.*** ***@***.***>

**Comment by mbezuidenhout** [comment]
See the file in /tasmota/tasmota_xsns_sensor/xsns_87_mcp2515.ino
I see there is a good article on how to communicate with that power supply here [https://www.beyondlogic.org/review-huawei-r4850g2-power-supply-53-5vdc-3kw/](https://www.beyondlogic.org/review-huawei-r4850g2-power-supply-53-5vdc-3kw/)

**Comment by reset12** [comment]
Hello Marius,
the beyoundlogic source is the reason why i like to use this power supply.
There is another guy on Youtube (Bavariansuperguy) which has for ESP a Beta-Software wich i cant load with ESPtool because of an error. I have all my devices working on tasmota so i like to use an ESP instead a Raspberry. I think, my problem can be the compiled bin, because i cant find the parameters i got

* CAN sniffer using MCP2515 - Microchip CAN controller
*
* Connections:
* MCP2515  ESP8266         Tasmota
* -------  --------------  ----------
*  INT     not used
*  SCK     GPIO14          SPI CLK
*  SI      GPIO13          SPI MOSI
*  SO      GPIO12          SPI MISO
*  CS      GPIO0..5,15,16  MCP2515
*  Gnd     Gnd
*  VCC     Vin/5V

The only selections i have can you see in the screeshots.



I have also tried the tasmots-sensors.bin but it seems als not work. With tasmota-sensors.bin i have success with ADS1115.

Do you have a Screeshot of Config module and Main Screen of a runnig system with MCP2515 ?

with regards
Bruno

On Thu, 02 Mar 2023 05:31:47 -0800
Marius Bezuidenhout ***@***.***> wrote:

> See the file in /tasmota/tasmota_xsns_sensor/xsns_87_mcp2515.ino
> I see there is a good article on how to communicate with that power supply here [https://www.beyondlogic.org/review-huawei-r4850g2-power-supply-53-5vdc-3kw/](https://www.beyondlogic.org/review-huawei-r4850g2-power-supply-53-5vdc-3kw/)
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/arendst/Tasmota/discussions/18084#discussioncomment-5178813
> You are receiving this because you authored the thread.
>
> Message ID: ***@***.***>


--
***@***.*** ***@***.***>

**Reply by barbudor** [reply]
@reset12 your compile must also have `#define USE_SPI` in addition to `#define USE_MCP2515`
Do you want me to generate a bin ? Do you need anything more than those ?

Beside, as per mbezuidenhout this driver only support "FreedomWon batteries" (whatever it is 😄) so it is likely additional improvement to the driver is needed to work with your Huawei PSU.

**Comment by reset12** [comment]
It would be great if someone could update the driver for the Huawei R4859G2
, I'm overwhelmed.

The actual Status of the Project can you see on the Picture.





On Thu, 02 Mar 2023 12:33:49 -0800
Barbudor ***@***.***> wrote:

> @reset12 your compile must also have `#define USE_SPI` in addition to `#define USE_MCP2515`
> Do you want me to generate a bin ? Do you need anything more than those ?
>
> Beside, as per mbezuidenhout this driver only support "FreedomWon batteries" (whatever it is 😄) so it is likely additional improvement to the driver is needed to work with your Huawei PSU.
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/arendst/Tasmota/discussions/18084#discussioncomment-5183049
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>

--
***@***.*** ***@***.***>

**Comment by reset12** [comment]
![IMG_20230303_093547_104](https://user-images.githubusercontent.com/14790813/222673022-d9affc97-30e0-493d-a69d-7c39e8fbc589.jpg)

**Reply by barbudor** [reply]
It needs someone who own the device to update the driver
So as of today, this is likely you

**Comment by goodkiller** [comment]
Related with #18287
